### ISSUE TITLE
Automatic layer selection for gradient checkpointing based on memory budget

### ISSUE THREAD

**Issue by powerofaisinstudy-debug** [pytorch/pytorch#194696]
Right now if you want to use gradient checkpointing to fit a big model in limited GPU memory, you have to manually pick which layers to checkpoint. For anything that isn't a simple nn.Sequential stack, checkpoint_sequential doesn't really help since it just takes a segment count, not actual layer/memory info.

In practice this means a lot of trial and error — run training, hit OOM, go back, checkpoint a few more layers, rerun, repeat. Anyone training larger models on a single GPU or limited cluster has probably done this dance.

Would be great to have something like:

python
model = torch.utils.checkpoint.auto_checkpoint(model, memory_budget="8GB")

that profiles the model's memory usage per layer and picks the minimal set of checkpoints needed to fit under the budget, instead of the user guessing.

Rough idea for how it could work:

do a dry run (or use existing memory stats) to get per-layer activation memory
pick checkpoints greedily (or knapsack-style) to hit the memory target with least recompute
wrap those layers with the existing checkpoint() call

This would work for arbitrary architectures, not just sequential ones, which is the main limitation of checkpoint_sequential today.

Happy to help test this / sketch out the profiling logic if useful.

cc @soulitzer

**Comment by powerofaisinstudy-debug** [comment]
Hey @soultizer, I've put together an initial draft PR exploring an experimental `auto_checkpoint` utility to tackle this! 

You can check out the work-in-progress code here: https://github.com/powerofaisinstudy-debug/pytorch/pull/2

Would love any early thoughts or feedback on the direction!
